lib/puppet/transaction/report.rb

```diff
@@ -308,14 +308,21 @@ def initialize_from_hash(data)
     end
   end
 
-  def skip_or_to_data_hash(rs)
-    if rs.nil?
-      nil
-    elsif !rs.out_of_sync? && Puppet[:exclude_unchanged_resources]
-      {}
-    else
-      rs.to_data_hash
+  def resource_unchanged?(rs)
+    states = %w[failed failed_to_restart out_of_sync skipped]
```

Comment:
since the `resource_unchanged?` method gets called a lot, recommend creating a frozen constant for these states. Also recommend creating them as an array of symbols so that ruby doesn't implicitly convert the string to a symbol everytime `send` is called.